### PR TITLE
Histogram concatenation for a faster HistogramTransformer

### PR DESCRIPTION
**Pull request by sadiamhjbn**
Experiment with transform(X, average=True) to produce a mean embedding directly, instead of per-example embeddings, thus opening an opportunity of relying more heavily on numpy's speed.

### CODE REVIEW

File: qunfold/transformers.py [hunk 1]
```diff
@@ -133,22 +133,19 @@ def fit_transform(self, X, y):
...
-      )
-      fX.append(fX_j.toarray())
-    fX = np.stack(fX).swapaxes(0, 1).reshape((X.shape[0], -1))
-    if self.unit_scale:
-      fX = fX / fX.sum(axis=1, keepdims=True)
-    return fX
+    return self._transform_after_preprocessor(X, average=average)
+  def _transform_after_preprocessor(self, X, average=False):
+    if not average:
+        return super()._transform_after_preprocessor(X) # FIXME
```
Comment (mirkobunse) [inline qunfold/transformers.py:142]: At this point, `super()._transform_after_preprocessor` does not exist (it only existed in the performance test script where you implemented a subclass of the HistogramTransformer; here, the change affects the HistogramTransformer directly, so that `super()` is a different object).

Instead of  `super()._transform_after_preprocessor`, you need to put the original code of `_transform_after_preprocessor`, starting at the line with `fX = []` until `return fX`.

File: qunfold/transformers.py [hunk 2]
```diff
@@ -133,22 +133,19 @@ def fit_transform(self, X, y):
...
+    # TODO: return concatenation of numpy histograms
```
Comment (mirkobunse) [inline qunfold/transformers.py:143]: please remove the "TODO: " statement because this task is done by now

File: qunfold/transformers.py (hunk 1 above)
Comment (mirkobunse) [reply qunfold/transformers.py:142]: Once you have made this change locally, please check that the unit tests of this package succeed. How to start the unit tests is described [in the documentation](https://mirkobunse.github.io/qunfold/developer-guide.html#unit-testing).

**Review by mirkobunse — CHANGES_REQUESTED** [review]
Hi Sadia, thanks for creating this pull request! I have added comments to places where changes need to be implemented.

File: qunfold/transformers.py (hunk 1 above)
Comment (sadiamhjbn) [reply qunfold/transformers.py]: Thanks for your feedback. I have edited the code as you suggested and run the unit test. I have also committed the edited code. Please let me know your opinion.

**Comment by mirkobunse** [comment]
Hi Sadia,

I've updated `qunfold/tests/performance_HDx.py` to check the correctness of this implementation. It turns out that `transform(X, average=True)` does often not produce the same result as `transform(X).mean(axis=0)`.

I've also found the origin of this error: You call `np.histogram(X[:, j], bins=self.n_bins)`, which results in bins that are different for each sample while the bins have to be identical for each sample. For this purpose, the HistogramTransformer creates `self.edges` during training. Please make sure that `np.histogram` uses these edges for all samples. Use my updated `qunfold/tests/performance_HDx.py` to check for correctness.

**Comment by mirkobunse** [comment]
Please check the documentation of `np.histogram` to see how specific edges can be supplied: https://numpy.org/doc/stable/reference/generated/numpy.histogram.html#numpy.histogram
